basic programs api setup and dashboard integration

zubair-arbi opened the pull request:
ECOM-2578
@awais786 @ahsan-ul-haq @tasawernawaz @afeef 
**WIP**

* Create common method `get_id_token` in `openedx/core/djangoapps/util/helpers.py` for use in `openedx -> programs` and `lms -> edxnotes`
* Create method `get_course_programs_for_dashboard` which gets `user` and `course_keys` as arguments and returns the `programs` dict.
* Add tests
* Update `edx-rest-api-client` version from `1.2.0` to `1.2.1` (Which includes new auth `SuppliedJwtAuth`)

_**Note:** The place for programs code, currently is `openedx/core/djangoapps/programs/`_

ahsan-ul-haq commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,84 @@
+        course_keys (list): List of course keys in which user is enrolled
+
+    Returns:
+        Json response containing programs or None
```
`JSON formatted programs containing given course_keys.`

ahsan-ul-haq commented on openedx/core/djangoapps/util/helpers.py:
```diff
@@ -0,0 +1,50 @@
+        String containing the signed JWT value or raise the exception
+        'ImproperlyConfigured'
+    """
+    import oauth2_provider.oidc as oidc
```
Add the `TODO` here instead of docstring.

zubair-arbi commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,83 @@
+
+    # get programs slumber-based client 'EdxRestApiClient'
+    try:
+        api_client = programs_api_client(ProgramsApiConfig.current().public_api_url, get_id_token(user, CLIENT_NAME))
```
@jimabramson I am using public api url for communication with programs. Is this ok?
`ProgramsApiConfig.current().public_api_url`

jimabramson replied:
@zubair-arbi because this call is between servers expected to reside on the same network, we should use the internal API URL.

zubair-arbi commented:
@rlucioni Please review and check if we need more logging.

zubair-arbi commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,83 @@
+        log.error('Failed to create programs api client.', exc_info=True)
+        return
+    except ImproperlyConfigured:
+        error_msg = "OAuth2 Client with name '{client_name}' is not present in the DB.".format(client_name=CLIENT_NAME)
```
@jimabramson @rlucioni I am catching exceptions and logging them so that the LMS dashboard doesn't crash.

zubair-arbi commented:
@edx/ecommerce

clintonb commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -0,0 +1,14 @@
+    """
+    return EdxRestApiClient(
+        api_url,
+        jwt=jwt_access_token
```
Why pass in the token when the client can generate it for you?

zubair-arbi replied:
We want to pass our own generated JWT for this we have utility method `get_id_token` (same as we do for edxnotes).
This was also the reason we implemented `SuppliedJwtAuth` in the PR: https://github.com/edx/edx-rest-api-client/pull/12

jimabramson commented on common/djangoapps/student/views.py:
```diff
@@ -573,6 +575,13 @@ def dashboard(request):
+    # pass this along in template context in order to render additional
+    # program-related information on the dashboard view.
+    course_programs = {}
+    if ProgramsApiConfig.current().is_student_dashboard_enabled:
```
i think it would be better to define a function in the programs djangoapp for  `is_student_dashboard_enabled` and have that function internally use the config model.   it's better if we do not import models across djangoapps.

jimabramson commented on lms/djangoapps/edxnotes/helpers.py:
```diff
@@ -21,15 +21,16 @@
 from opaque_keys.edx.keys import UsageKey
-from .exceptions import EdxNotesParseError, EdxNotesServiceUnavailable
+from edxnotes.exceptions import EdxNotesParseError, EdxNotesServiceUnavailable
+from openedx.core.djangoapps.util.helpers import get_id_token
```
since you're already making changes here, please fix the import organization.

jimabramson commented on openedx/core/djangoapps/programs/tests/test_models.py:
```diff
@@ -8,10 +8,9 @@
 
-@patch('config_models.models.cache.get', return_value=None)  # during tests, make every cache get a miss.
-class ProgramsApiConfigTest(TestCase):
+class ProgramsApiBasicTestConfig(TestCase):
```
I think this should be a mixin (`ProgramsApiConfigMixin`) instead of a base class, and reside in tests/mixins.py

jimabramson commented on openedx/core/djangoapps/programs/tests/test_views.py:
```diff
@@ -0,0 +1,98 @@
+    def setUp(self, **kwargs):  # pylint: disable=unused-argument
+        super(TestGetXSeriesPrograms, self).setUp()
+        self.create_config(enabled=True, enable_student_dashboard=True)
+        Client.objects.get_or_create(name="programs", client_type=0)
```
is there a constant we can use instead of the int literal for `client_type` ?

zubair-arbi replied:
I found it: https://github.com/caffeinehit/django-oauth2-provider/blob/master/provider/constants.py#L7
Its a tuple. So I think providing it like this is fine.

rlucioni replied:
@zubair-arbi you should be able to import and use `CONFIDENTIAL`.

jimabramson commented:
@zubair-arbi expected output from Programs API should be updated to include 'marketing_slug' since https://github.com/edx/programs/pull/19 has merged.

jimabramson commented on openedx/core/djangoapps/programs/tests/test_views.py:
```diff
@@ -0,0 +1,98 @@
+        """
+        # mock the request call
+        with patch('slumber.Resource.get') as mock_get:
+            mock_get.return_value = self.programs_api_response
```
If we ever called the API on behalf of an invalid or anonymous user, the response from the API would be a 401.  Also, I don't think it's important to test this function with an anonymous user, since they would never be looking at a student dashboard - is the reason for this test to ensure the exception handling is working?

zubair-arbi replied:
@jimabramson yes. Since I am catching and logging all exceptions to avoid any affect on student dashboard so that's why I am doing this.

jimabramson replied:
in that case what I recommend is to set `mock_get.side_effect` to `Exception` and rename the test method so it's clear you are testing for exception handling.

I also think it's worth another similar test method, in which you patch __init__ of the API client and make *that* raise an Exception, so both exception scenarios are covered.

zubair-arbi replied:
This one dosen't included in PR update. I will fix it.

jimabramson commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,83 @@
+
+To ensure maximum separation of concerns, and a minimum of interdependence's,
+this package should be kept small, thin, and stateless.
+"""
```
this docstring describes the package itself, so it should be in `__init__` or a README, not in the views module.

zubair-arbi replied:
Ok. I will move it over to __init__

jimabramson commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,83 @@
+        api_client = programs_api_client(ProgramsApiConfig.current().public_api_url, get_id_token(user, CLIENT_NAME))
+        log.info("Programs slumber-based client 'EdxRestApiClient' created successfully.")
+    except ValueError:
+        log.error('Failed to create programs api client.', exc_info=True)
```
N.B.:  `log.exception` is a nice shorthand for `log.error(exc_info=True)`

jimabramson commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,83 @@
+    except ImproperlyConfigured:
+        error_msg = "OAuth2 Client with name '{client_name}' is not present in the DB.".format(client_name=CLIENT_NAME)
+        log.error(error_msg, exc_info=True)
+        return
```
I don't think there's much value to catching specific kinds of exceptions, and we may end up making wrong assumptions about what went wrong.  I'd be more comfortable catching any `Exception` and dumping the trace without modification before returning.

@rlucioni this affects your work on log monitoring; do you agree with my recommendation?

jimabramson replied:
If we fail to create a client the function returns `None` but if we succeed and then fail to get an API response we return an empty dict.

Let's make this consistent and always return the empty dict, even if an error occurred.

rlucioni replied:
@jimabramson @zubair-arbi I'd also be more comfortable catching any exception and logging the trace. Not only does it make for cleaner code here, it should facilitate debugging down the road. I'd rather investigate the cause of errors during debugging than be presented with a potentially incorrect assumption.

rlucioni replied:
@zubair-arbi I would just do something like the following:
```python
try:
    api_client = programs_api_client(ProgramsApiConfig.current().public_api_url, get_id_token(user, CLIENT_NAME))
except:
    log.exception('Failed to initialize Programs API client.')
```

jimabramson commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,83 @@
+        return course_programs
+
+    programs = response.get('results')
+    if programs is None:
```
`None` isn't an expected response from the API, it should always be a list (that may be empty).

zubair-arbi replied:
You are right. I am implementing this and it will always be a dictionary.

jimabramson commented on openedx/core/djangoapps/util/helpers.py:
```diff
@@ -0,0 +1,48 @@
+        String containing the signed JWT value or raise the exception
+        'ImproperlyConfigured'
+    """
+    # TODO: there's a circular import problem somewhere which is why we do the oidc import inside of this function.
```
is this still a problem, after moving this function to this djangoapp?

zubair-arbi replied:
I don't think this is still a problem, I was waiting till I fix the `edxnotes` tests. I will update this.

zubair-arbi replied:
@jimabramson `ally` tests still failing due to circular import. The full traceback is here: https://gist.github.com/zubair-arbi/bec778d5d3e7c0ca79ac

jimabramson replied:
fixing this issue should be a separate ticket as well.   I'm ok with the internal import for now.

jimabramson commented on openedx/core/djangoapps/util/helpers.py:
```diff
@@ -0,0 +1,48 @@
+log = logging.getLogger(__name__)
+
+
+def get_id_token(user, client_name):
```
Was there a test left behind in edxnotes for this?  If so it needs to come along.  If not, please create a JIRA ticket to add tests for this function.

zubair-arbi replied:
I will create a new Jira ticket.

jimabramson commented on openedx/core/djangoapps/programs/tests/test_views.py:
```diff
@@ -0,0 +1,98 @@
+                    'name': 'First Program'
+                }
+            }
+            self.assertEqual(expected_output, programs)
```
for better coverage of the logic, please add a test for the cases where:
* each of the two different programs matches one of the courses
* one of the programs matches both of the courses

rlucioni commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,83 @@
+supporting learner- and author-facing features involving that service
+if and only if the service is deployed in the Open edX installation.
+
+To ensure maximum separation of concerns, and a minimum of interdependence's,
```
s/interdependence's/interdependencies

rlucioni commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,83 @@
+    try:
+        response = api_client.programs.get()
+    except Exception:  # pylint: disable=broad-except
+        log.error('Failed to get programs from api client.', exc_info=True)
```
This could be:
```python
try:
    response = api_client.programs.get()
except:
    log.exception('Failed to retrieve programs.')
```

rlucioni commented on openedx/core/djangoapps/util/helpers.py:
```diff
@@ -0,0 +1,48 @@
+        access_token = access_tokens[0]
+    else:
+        access_token = AccessToken(client=client, user=user)
+        access_token.save()
```
You can fold this into one line with `access_token = AccessToken.create(client=client, user=user)`

zubair-arbi replied:
exception:
`AttributeError: type object 'AccessToken' has no attribute 'create'`

rlucioni replied:
`AccessToken.objects.create(client=client, user=user)` is what you want.

rlucioni commented on openedx/core/djangoapps/util/helpers.py:
```diff
@@ -0,0 +1,48 @@
+        client=client,
+        user__username=user.username,
+        expires__gt=now()
+    ).order_by('-expires')
```
Consider using `first()` to simplify the below:
```python
access_token = AccessToken.objects.filter(
    client=client,
    user__username=user.username,
    expires__gt=now()
).order_by('-expires').first()

if not access_token:
    # Create access token
```

zubair-arbi replied:
I think its fine for now, anyways I am getting following exception:
`AttributeError: 'QuerySet' object has no attribute 'first'`

rlucioni replied:
It's not an attribute, it's a method. Using it would make for more idiomatic use of Django's `QuerySet`.

zubair-arbi replied:
You are right. But my test is failing with the above exception.

jimabramson replied:
that is strange, it seems that it should work per [the documentation](https://docs.djangoproject.com/en/1.8/ref/models/querysets/#django.db.models.query.QuerySet.first).  However, as it's mainly a convenience, I'm OK with bypassing it for now if the root cause is elusive.

zubair-arbi commented:
@rlucioni @jimabramson other the one or two comments I have addressed all comments/suggestions. PR is now ready for another review.

jimabramson commented on openedx/core/djangoapps/programs/tests/test_views.py:
```diff
@@ -0,0 +1,135 @@
+            ]
+        }
+
+    def test_get_course_programs_with_valid_user_and_courses(self):
```
in addition to this, i still think it's important to have a test method where there are two different programs, to make sure the function is assigning programs to courses correctly.  you can mostly reuse stuff you already have in this test, but need a second example response program, and slightly different inputs to the get_course_programs function.

zubair-arbi replied:
I have removed this test and created a new one with the name `test_get_course_programs_with_exception`

zubair-arbi replied:
Sorry, in above comment I wanted to tell about the old test for anonymous user.

jimabramson commented on openedx/core/djangoapps/programs/tests/test_views.py:
```diff
@@ -0,0 +1,135 @@
+        """
+        # mock the request call
+        with patch('slumber.Resource.get') as mock_get:
+            mock_get.return_value = self.programs_api_response
```
as discussed earlier, instead of setting return_value, set side_effect to Exception, since a JSON response to an anonymous user is not a realistic scenario.

jimabramson commented on openedx/core/djangoapps/programs/tests/test_views.py:
```diff
@@ -0,0 +1,135 @@
+
+
+@skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Test only valid in lms')
+class TestGetXSeriesPrograms(ProgramsApiConfigMixin, TestCase):
```
in the test methods in this TestCase that use mock_get, please add assertions that mock_get is actually being called...

zubair-arbi commented:
@jimabramson Please review again.

jimabramson commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,68 @@
+    """
+    course_programs = {}
+    if not ProgramsApiConfig.current().is_student_dashboard_enabled:
+        log.info("Programs service for student dashboard is disabled.")
```
It's probably best to make this `log.warning` instead of `info` because in practice, this function should not even be getting called if not configured (there should be a check before calling it on the dashboard, which I see in #10372), but I like that we're being extra defensive here.

jimabramson commented on openedx/core/djangoapps/programs/tests/test_views.py:
```diff
@@ -0,0 +1,214 @@
+        # mock the request call
+        with patch('slumber.Resource.get') as mock_get:
+            # test output in case of the exception 'ValueError'
+            mock_get.side_effect = ValueError('exc')
```
I would be comfortable with a single test here for `Exception` instead of one each for (`ValueError`, `ImproperlyConfigured`, `HttpClientError`).  However, I think we should also have a test where we mock an exception coming from patching `programs_api_client` (or `slumber.API.__init__`, either way is equivalent I think) - this will exercise the other place in which we are catching exceptions in that function.

jimabramson commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,68 @@
+
+    programs = response.get('results', [])
+    if not programs:
+        log.info("No programs found for the user '%s'.", user.username)
```
this is also an unexpected case, so `log.warning` makes more sense to me here.

jimabramson commented:
@zubair-arbi it looks like the python tests that are failing can be resolved by updating the `@patch` directives in https://github.com/edx/edx-platform/blob/master/lms/djangoapps/edxnotes/tests.py that still expect `get_id_token` to be found in the original location.

jimabramson commented:
@zubair-arbi also, these are the tests that I can find for `get_id_token`: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/edxnotes/tests.py#L959-L977  I think those tests can be left in place, because they are exercising a URL in edxnotes as well as in the token function.

I think it would make sense to create a new test case in a `test_helpers.py` module which is similar to `test_get_id_token` in the edxnotes module.  It should ensure that we can `jwt.decode` the result of the function, and that it works for each of two scenarios: user has an access token already, user has no access token already.

jimabramson commented:
@zubair-arbi aside from logging and test-related modifications i've requested, this is looking good!

zubair-arbi commented:
@jimabramson I have addressed your comments and added the test for method `get_id_token`.
There is one issue, I am unable to path the method `get_id_token` from `openedx` into `lms -> edxnotes`.

zubair-arbi commented:
@jimabramson I did some refactoring for `edxnotes` and I think that the failing tests are fixed. I am waiting for build results now.

zubair-arbi commented:
@jimabramson @rlucioni PR is ready for final review.

jimabramson commented on common/djangoapps/student/views.py:
```diff
@@ -573,6 +574,13 @@ def dashboard(request):
+    # pass this along in template context in order to render additional
+    # program-related information on the dashboard view.
+    course_programs = {}
+    if is_student_dashboard_programs_enabled:
```
missing `()`

rlucioni commented on lms/djangoapps/edxnotes/helpers.py:
```diff
@@ -1,35 +1,39 @@
 import logging
 import requests
+
+from datetime import datetime
```
Minor: since you're making changes here, please fix the import order. Requests is a third-party library; datetime, json, and uuid are part of the standard library.

rlucioni commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,68 @@
+    try:
+        response = api_client.programs.get()
+    except Exception:  # pylint: disable=broad-except
+        log.exception('Failed to retrieve xseries programs from the Programs API client.')
```
Minor, but calling the Programs API at this point will return all programs. "Failed to retrieve programs from the Programs API" might be more accurate.

rlucioni commented on openedx/core/djangoapps/programs/views.py:
```diff
@@ -0,0 +1,68 @@
+
+    programs = response.get('results', [])
+    if not programs:
+        log.warning("No programs found for the user '%s'.", user.username)
```
We try to keep usernames out of the logs when possible. Could you please log the user's ID instead?

zubair-arbi replied:
Sure.

rlucioni commented:
@zubair-arbi I made a few minor comments, but :+1: otherwise. Thank you for your hard work here.

jimabramson commented on openedx/core/djangoapps/programs/tests/test_views.py:
```diff
@@ -0,0 +1,208 @@
+        based client 'programs_api_client'.
+        """
+        # mock the request call
+        with patch('edx_rest_api_client.client.EdxRestApiClient.__init__') as mock_get:
```
it would clearer to name the patch variable `mock_init` here

jimabramson commented on openedx/core/djangoapps/programs/tests/test_views.py:
```diff
@@ -0,0 +1,208 @@
+        # mock the request call
+        with patch('edx_rest_api_client.client.EdxRestApiClient.__init__') as mock_get:
+            # test output in case of the exception 'ValueError'
+            mock_get.side_effect = ValueError('exc')
```
if we are testing that any Exception will be handled, then the side effect should be `Exception` instead of `ValueError`

jimabramson commented on openedx/core/djangoapps/programs/tests/test_views.py:
```diff
@@ -0,0 +1,208 @@
+        """
+        # mock the request call
+        with patch('slumber.Resource.get') as mock_get:
+            mock_get.side_effect = HttpClientError('exc')
```
s/`HttpClientError`/`Exception`/

jimabramson commented:
@zubair-arbi I have requested a few more small changes to tests, but once those are fixed (and commits squashed) :+1:

zubair-arbi commented:
@rlucioni @jimabramson I have addressed your comment and squashed the commit. Now waiting for build to pass before merging it into master.

jimabramson commented:
jenkins run bokchoy

jimabramson commented:
@zubair-arbi i think the failure is unrelated but I don't see a flaky_test ticket for it so running bokchoy again to see what happens

zubair-arbi commented:
@jimabramson all the tests passed.

jimabramson commented:
woohoo!   :ship:
